### PR TITLE
Use ThreadMode to replace explicit thread number

### PR DESCRIPTION
**Pull request by upsuper**
There are several advantages to use a thread mode instead of explicit thread number:
1. all values of thread mode are sensible, while only restricted values of thread number makes sense
2. errors for thread number are no longer necessary

### CODE REVIEW

File: src/argon2.rs [hunk 1]
```diff
@@ -502,7 +495,7 @@ pub fn verify_raw_std(
                mem_cost,
                time_cost,
                parallelism,
-               parallelism,
+               ThreadMode::Parallel,
```
Comment (mrijkeboer) [inline src/argon2.rs]: If parallelism is 1 this should be ThreadMode::Sequential.

File: src/encoding.rs [hunk 2]
```diff
@@ -380,7 +381,7 @@ mod tests {
                                    mem_cost,
                                    time_cost,
                                    parallelism,
-                                   parallelism,
+                                   ThreadMode::Parallel,
```
Comment (mrijkeboer) [inline src/encoding.rs]: If parallelism is 1 this should be ThreadMode::Sequential.

**Comment by upsuper** [comment]
I think it makes more sense to have ThreadMode's default value to be parallel, and check in core that if lanes is 1, use sequential path whatever the mode is. 

I want to propose another change which moves all the basic info (variant, version, lanes, thread mode, costs) into a Config struct, so that we don't need separate functions for "default" and non-default. User can easily generate a default config via Default::default(), and they can specify part of config with rest of them being the default value.

With that change, it makes sense to have it run in parallel mode by default if there are more than one lanes.

**Comment by mrijkeboer** [comment]
> I think it makes more sense to have ThreadMode's default value to be parallel, and check in core that if lanes is 1, use sequential path whatever the mode is.

I chose sequential as default value, because the [original C version](https://github.com/P-H-C/phc-winner-argon2) has this as default and I want my Rust implementation to match the original as close as possible. Furthermore, the thread mode must be separate from lanes, since the lanes value is part of the hash and therefore can't be changed afterwards. For instance, if you have a hash that was created with 4 lanes and you want to validate that hash on a system that can't do threads you have to set the Threadmode to sequential, if you would alter the number of lanes the hash wouldn't match.

> I want to propose another change which moves all the basic info (variant, version, lanes, thread mode, costs) into a Config struct, so that we don't need separate functions for "default" and non-default. User can easily generate a default config via Default::default(), and they can specify part of config with rest of them being the default value.

Initially I thought about this too (I saw the HashOptions in the [Haskell implementation](https://hackage.haskell.org/package/argon2)) however, I created this crate with the idea to export its functionality to different languages using FFI and a configuration struct would make that more involved. However, the plan has changed and I'm currently writing a separate crate to export the functionality using FFI, therefore a configuration struct would no longer be a problem. Furthermore, since you brought it up, it seems to be a useful feature and I'm even willing to write it myself if you haven't written it already. In case you have written it already, I would appreciate a separate pull request.

**Comment by upsuper** [comment]
> Furthermore, since you brought it up, it seems to be a useful feature and I'm even willing to write it myself if you haven't written it already. In case you have written it already, I would appreciate a separate pull request.

I haven't. Great that you are going to implement that. Thanks for the work.

And Ok, I think those reasons make sense, and I'll update the pr to use sequential by default.

**Comment by upsuper** [comment]
Updated with the default value set to sequential, and using sequential function when lanes == 1.

**Comment by mrijkeboer** [comment]
There're some places where the ThreadMode handling is different than the original parallelism handling, but those will be replaced with the upcoming Config change so I'll merge it anyway. Thanks for your work.

**Pull request merged**